libexec/getty/ttys.5
.\" Copyright (c) 1985, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)ttys.5	8.1 (Berkeley) 6/4/93
.\" "
.Dd August 3, 2018
.Dt TTYS 5
.Os
.Sh NAME
.Nm ttys
.Nd terminal initialization information
.Sh DESCRIPTION
The file
.Nm
contains information that is used by various routines to initialize
and control the use of terminal special files.
Pseudo-terminals (see
.Xr pts 4 )
are not listed.
This information is read with the
.Xr getttyent 3
library routines.
There is one line in the
.Nm
file per special device file.
Fields are separated by tabs and/or spaces.
Fields comprised of more than one word should be enclosed in double
quotes (``"'').
Blank lines and comments may appear anywhere in the file; comments
are delimited by hash marks (``#'') and new lines.
Any unspecified fields will default to null.
.Pp
The first field is normally the
name of the terminal special file as it is found in
.Pa /dev .
However, it can be any arbitrary string
when the associated command is not related to a tty.
.Pp
The second field of the file is the command to execute for the line,
usually
.Xr getty 8 ,
which initializes and opens the line, setting the speed, waiting for
a user name and executing the
.Xr login 1
program.
It can be, however, any desired command, for example
the start up for a window system terminal emulator or some other
daemon process, and can contain multiple words if quoted.
.Pp
The third field is the type of terminal usually connected to that
tty line, normally the one found in the
.Xr termcap 5
data base file.
The environment variable
.Ev TERM
is initialized with the value by
either
.Xr getty 8
or
.Xr login 1 .
.Pp
The remaining fields set flags in the
.Fa ty_status
entry (see
.Xr getttyent 3 ) ,
specify a window system process that
.Xr init 8
will maintain for the terminal line, optionally determine the
type of tty (whether dialin, network or otherwise),
or specify a tty group
name that allows the login class database (see
.Xr login.conf 5 )
to refer to many ttys as a group, to selectively allow or
deny access or enable or disable accounting facilities for
ttys as a group.
.Pp
As flag values, the strings ``on'' and ``off'' specify that
.Xr init 8
should (should not) execute the command given in the second field.
``onifconsole'' will cause this line to be enabled if and only if it is
an active kernel console device (it is equivalent to ``on'' in this
case).
The flag ``onifexists'' will cause this line to be enabled if and only
if the name exists.
If the name starts with a ``/'', it will be considered an absolute
path.
Otherwise, it is considered a path relative to
.Pa /dev .
The flag ``secure'' (if the console is enabled) allows users with a
uid of 0 to login on
this line.
The flag ``dialin'' indicates that a tty entry describes a dialin
line, and ``network'' is obsolete and does nothing.
Either of these strings may also be specified in the terminal type
field.
The string ``window='' may be followed by a quoted command
string which
.Xr init 8
will execute
.Em before
starting the command specified by the second field.
.Pp
The string ``group='' may be followed by a group name comprised of
alphanumeric characters that can be used by
.Xr login.conf 5
to refer to many tty lines as a group to enable or disable access
and accounting facilities.
If no group is specified, then the tty becomes a member of the group
"none".
For backwards compatibility, the ``group='' should appear last on the
line, immediately before the optional comment.
.Pp
Both the second field and any command specified with ``window=''
will be split into words and executed using
.Xr execve 2 .
Words are separated by any combinations of tabs and spaces.
Arguments containing whitespace should be enclosed in single quotes
.Pq Li ' .
Note that no shell-style globbing or other variable substitution occurs.
.Sh FILES
.Bl -tag -width /etc/ttys -compact
.It Pa /etc/ttys
.El
.Sh EXAMPLES
.Bd -literal
# root login on console at 1200 baud
console	"/usr/libexec/getty std.1200"	vt100	on secure
# dialup at 1200 baud, no root logins
ttyd0	"/usr/libexec/getty d1200"	dialup	on group=dialup	# 555-1234
# Mike's terminal: hp2621
ttyh0	"/usr/libexec/getty std.9600"	hp2621-nl	on group=dialup	# 457 Evans
# John's terminal: vt100
ttyh1	"/usr/libexec/getty std.9600"	vt100	on group=dialup		# 459 Evans
# terminal emulate/window system
ttyv0	"/usr/local/bin/xterm -display :0"	xterm	on window="/usr/local/bin/X :0"
.Ed
.Sh SEE ALSO
.Xr login 1 ,
.Xr getttyent 3 ,
.Xr gettytab 5 ,
.Xr login.conf 5 ,
.Xr termcap 5 ,
.Xr getty 8 ,
.Xr init 8 ,
.Xr pam_securetty 8 ,
.Xr pstat 8
.Sh HISTORY
A
.Nm
file appeared in
.At v6 .


NEWS
GNU diffutils NEWS                                    -*- outline -*-

* Noteworthy changes in release 3.2 (2011-09-02) [stable]

** Changes in behavior

  --ignore-file-name-case now applies at the top level too.
  For example, "diff dir inIt" might compare "dir/Init" to "inIt".

** New features

  diff and sdiff have a new option --ignore-trailing-space (-Z).

** Packaging

  The texinfo documentation no longer specifies "front-cover" or "back-cover"
  texts, so that it may now be included in Debian's "main" section.


* Noteworthy changes in release 3.1 (2011-08-10) [stable]

** Bug fixes

  diff no longer reports spurious differences merely because two entries
  in the same directory have names that compare equal in the current
  locale, or compare equal because --ignore-file-name-case was given.

* Noteworthy changes in release 3.0 (2010-05-03) [stable]

** Bug fixes

  diff once again prints the required "\ No newline at end of file" line
  when at least one input lacks a newline-at-EOF and the final hunk plus
  context-length aligns exactly with the end of the newline-lacking file.
  [bug introduced between 2.8.7 and 2.9]

** Changes in behavior

  In context-style diffs, diff prints a portion of a preceding "function"
  line for each hunk, with --show-function-line=RE (-F) or
  --show-c-function (-p).  Now, it trims leading blanks from such lines
  before extracting a prefix.  This is useful especially when a function
  line is so far indented that the name itself would be truncated or not
  included in the limited-width substring that diff appends.

  diff once again reports a difference with the diagnostic
  "Binary files A and B differ" when at least one of the files
  appears to be binary.  From 2.8.4 through diffutils-2.9, it printed
  "Files A and B differ".


* Noteworthy changes in release 2.9 (2010-02-11) [stable]

** New features

  New diff option --suppress-blank-empty.

  Bring back support for `diff -NUM', where NUM is a number,
  even when conforming to POSIX 1003.1-2001.  This change reverts to
  the behavior of GNU diff 2.7 and earlier.  This is a change only
  when conforming to POSIX 1003.1-2001; there is no effect when
  conforming to older POSIX versions.

  This change is in response to decisions taken in the January 2005
  Austin Group standardization meeting.  For more details, please see
  "Utility Syntax Guidelines" in the Minutes of the January 2005
  Meeting <http://www.opengroup.org/austin/docs/austin_239.html>.

  sdiff now understands '1' and '2' as synonyms for 'l' and 'r'.

** Changes in behavior

  sdiff and diff3 now invoke diff, not $(bindir)/diff

** Administrivia

  New discussion and bug-reporting address: bug-diffutils@gnu.org

  updated gnulib support


* Noteworthy changes in release 2.8.7 (2004-04-13) [stable]

  Version 2.8.7 contains no user-visible changes.

User-visible changes in version 2.8.6:

* New diff3 option --strip-trailing-cr.

* With -N and -P, inaccessible empty regular files (the kind of files
  that 'patch' creates to indicate nonexistent backups) are now
  treated as nonexistent when they are in the 'backup' file position.

* If multiple SKIP values are given to cmp, e.g., `cmp -i 10 -i 20',
  cmp now uses the maximal value instead of the last one.

* diff now omits the ".000000000" on hosts that do not support
  fractional time stamps.

Version 2.8.5 was not publicly released.

User-visible changes in version 2.8.4:

* Diff now simply prints "Files A and B differ" instead of "Binary
  files A and B differ".  The message is output if either A or B
  appears to be a binary file, and the old wording was misleading
  because it implied that both files are binary, which is not
  necessarily the case.

User-visible changes in version 2.8.3:

* New locale: en_US.

User-visible changes in version 2.8.2:

* New diff and sdiff option:
  --tabsize=COLUMNS
* If --ignore-space-change or --ignore-all-space is also specified,
  --ignore-blank-lines now considers lines to be empty if they contain
  only white space.
* More platforms now handle multibyte characters correctly when
  excluding files by name (diff -x and -X).
* New locales: hu, pt_BR.

User-visible changes in version 2.8.1:

* Documentation fixes.

User-visible changes in version 2.8:

* cmp and diff now conform to POSIX 1003.1-2001 (IEEE Std 1003.1-2001)
  if the underlying system conforms to POSIX and if the _POSIX2_VERSION
  environment variable is set to 200112.  Conformance removes support
  for `diff -NUM', where NUM is a number.  Use -C NUM or -U NUM instead.
* cmp now supports trailing operands SKIP1 and SKIP2, like BSD cmp.
* cmp -i or --ignore-initial now accepts SKIP1:SKIP2 option value.
* New cmp option: -n or --bytes.
* cmp's old -c or --print-chars option has been renamed;
  use -b or --print-bytes instead.
* cmp now outputs "byte" rather than "char" outside the POSIX locale.
* cmp -l's index column width now adjusts to fit larger (or smaller) files.
* cmp -l -s and cmp -s -l are not allowed.  Use cmp -s or cmp -l instead.
* diff uses ISO 8601 style time stamps for output times (e.g. "2001-11-23
  16:44:36.875702460 -0800") unless in the C or POSIX locale and the
  -c style is specified.
* diff's -I and -F options use the regexp syntax of grep, not of Emacs.
* diff now accepts multiple context arguments, and uses their maximum value.
* New diff and sdiff options:
  -E  --ignore-tab-expansion
  --strip-trailing-cr
* New diff options:
  --from-file=FILE, --to-file=FILE
  --ignore-file-name-case
  --no-ignore-file-name-case
* New diff3 and sdiff option:
  --diff-program=PROGRAM
* The following diff options are still accepted, but are no longer documented.
  They may be withdrawn in future releases.
  -h (omit; it has no effect)
  -H (use --speed-large-files instead)
  -L (use --label instead)
  -P (use --unidirectional-new-file instead)
  --inhibit-hunk-merge (omit; it has no effect)
* Recursive diffs now sort file names according to the LC_COLLATE locale

usr.sbin/sysinstall/ttys.c
/*
 * The new sysinstall program.
 *
 * This is probably the last program in the `sysinstall' line - the next
 * generation being essentially a complete rewrite.
 *
 * $FreeBSD: stable/9/usr.sbin/sysinstall/ttys.c 83845 2001-09-22 23:17:37Z murray $
 *
 * Copyright (c) 2001
 *      Andrey A. Chernov.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer,
 *    verbatim and that no modifications are made prior to this
 *    point in the file.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY ANDREY A. CHERNOV ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL ANDREY A. CHERNOV OR HIS PETS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, LIFE OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 */

#include "sysinstall.h"
#include <sys/stat.h>
#include <ctype.h>
#include <ttyent.h>

#define _X_EXTENSION ".XXXXXX"

void
configTtys(void)
{
    size_t len;
    int t, tlen, changed;
    FILE *fp, *np;
    char sq, *line, *p, *q, *cp, *tptr;
    char templ[sizeof(_PATH_TTYS) + sizeof(_X_EXTENSION) - 1];
    struct ttyent *tnam;
    
    if ((cp = variable_get(VAR_CONSTERM)) == NULL ||
	strcmp(cp, "NO") == 0)
	return;
    if (!file_readable(_PATH_TTYS)) {
	msgConfirm("%s not exist or not readable", _PATH_TTYS);
	return;
    }
    if ((fp = fopen(_PATH_TTYS, "r")) == NULL) {
	msgConfirm("Can't open %s for read: %s", _PATH_TTYS,
		   strerror(errno));
	return;
    }
    strcpy(templ, _PATH_TTYS _X_EXTENSION);
    if ((t = mkstemp(templ)) < 0) {
	msgConfirm("Can't create %s: %s", templ, strerror(errno));
	(void)fclose(fp);
	return;
    }
    if (fchmod(t, 0644)) {
	msgConfirm("Can't fchmod %s: %s", templ, strerror(errno));
	(void)fclose(fp);
	return;
    }
    if ((np = fdopen(t, "w")) == NULL) {
	msgConfirm("Can't fdopen %s: %s", templ, strerror(errno));
	(void)close(t);
	(void)fclose(fp);
	(void)unlink(templ);
	return;
    }
    changed = 0;
    while ((line = fgetln(fp, &len)) != NULL) {
	p = line;
	while (p < (line + len) && isspace((unsigned char)*p))
	    ++p;
	if (strncmp(p, "ttyv", 4) != 0) {
    dump:
	    if (fwrite(line, len, 1, np) != 1) {
    wrerr:
		msgConfirm("%s: write error: %s", templ, strerror(errno));
		(void)fclose(fp);
		(void)fclose(np);
		(void)unlink(templ);
		return;
	    }
	} else {
	    q = p;
	    while(q < (line + len) && !isspace((unsigned char)*q))
		++q;
	    if (!isspace((unsigned char)*q))
		goto dump;
	    sq = *q;
	    *q = '\0';
	    tnam = getttynam(p);
	    *q = sq;
	    if (tnam == NULL || tnam->ty_type == NULL ||
		strcmp(tnam->ty_type, cp) == 0 ||
		strncmp(tnam->ty_type, "cons", 4) != 0 ||
		!isdigit((unsigned char)tnam->ty_type[4])
	       )
		goto dump;
	    tlen = strlen(tnam->ty_type);
	    tptr = NULL;
	    p = ++q;
	    while(p < (line + len)) {
		if (strncmp(p, tnam->ty_type, tlen) == 0) {
		    tptr = p;
		    break;
		}
		++p;
	    }
	    if (tptr == NULL)
		goto dump;
	    changed = 1;
	    if (fwrite(line, tptr - line, 1, np) != 1 ||
		fputs(cp, np) ||
		fwrite(tptr + tlen,
		       len - (tptr + tlen - line), 1, np) != 1)
		goto wrerr;
	}
    }
    if (!feof(fp)) {
	msgConfirm("%s: read error: %s", _PATH_TTYS, strerror(errno));
	(void)fclose(fp);
	(void)fclose(np);
	(void)unlink(templ);
	return;
    }
    (void)fclose(fp);
    if (fclose(np)) {
	if (changed)
		msgConfirm("%s: close error: %s", templ, strerror(errno));
	else
		variable_set2(VAR_CONSTERM, "NO", 0);
	(void)unlink(templ);
	return;
    }
    if (!changed) {
	(void)unlink(templ);
	variable_set2(VAR_CONSTERM, "NO", 0);
	return;
    }
    if (rename(templ, _PATH_TTYS)) {
	msgConfirm("Can't rename %s to %s: %s", templ, _PATH_TTYS,
		   strerror(errno));
	return;
    }
    variable_set2(VAR_CONSTERM, "NO", 0);
}
